src/com/google/enterprise/adaptor/database/DatabaseAdaptor.java

```diff
@@ -633,25 +672,73 @@ private void addMetadata(MetadataHandler meta, ResultSet rs)
   @VisibleForTesting
   void addMetadataToRecordBuilder(final DocIdPusher.Record.Builder builder,
       ResultSet rs) throws SQLException, IOException {
+    synchronized (this) {
+      if (metadataColumns == null) {
+        metadataColumns = new MetadataColumns(rs.getMetaData());
+      }
+    }
     addMetadata(
         new MetadataHandler() {
           @Override public void addMetadata(String k, String v) {
             builder.addMetadata(k, v);
           }
         },
-        rs);
+        rs, metadataColumns);
   }
 
   @VisibleForTesting
   void addMetadataToResponse(final Response resp, ResultSet rs)
       throws SQLException, IOException {
+    synchronized (this) {
+      if (metadataColumns == null) {
+        metadataColumns = new MetadataColumns(rs.getMetaData());
+      }
+    }
     addMetadata(
         new MetadataHandler() {
           @Override public void addMetadata(String k, String v) {
             resp.addMetadata(k, v);
           }
         },
-        rs);
+        rs, metadataColumns);
+  }
+
+  private void addExtraMetadataToRecordBuilder(
+      final DocIdPusher.Record.Builder builder, Connection conn,
+      String uniqueId) throws SQLException, IOException {
+    MetadataHandler handler = new MetadataHandler() {
+        @Override public void addMetadata(String k, String v) {
+          builder.addMetadata(k, v);
+        }
+      };
+    addExtraMetadata(conn, uniqueId, handler);
+  }
+
+  private void addExtraMetadataToResponse(final Response resp, Connection conn,
+      String uniqueId) throws SQLException, IOException {
+    MetadataHandler handler = new MetadataHandler() {
+        @Override public void addMetadata(String k, String v) {
+          resp.addMetadata(k, v);
+        }
+      };
+    addExtraMetadata(conn, uniqueId, handler);
+  }
+
+  private void addExtraMetadata(Connection conn, String uniqueId,
+      MetadataHandler handler) throws SQLException, IOException {
+    try (PreparedStatement stmt = prepareExtraMetadataStatement(conn, uniqueId);
+        ResultSet rs = stmt.executeQuery()) {
+      synchronized (this) {
+        if (extraMetadataColumns == null) {
+          extraMetadataColumns = new MetadataColumns(rs.getMetaData());
+        }
+      }
+      while (rs.next()) {
+        addMetadata(handler, rs, extraMetadataColumns);
+      }
+    } catch (SQLException ex) {
+      throw new IOException(ex);
```

Comment:
This line is untested. I'm not sure how to trigger this. A flat invalid query would fail in validation. Maybe a type mismatch in the parameter expression: "<int-field> = ?" with a character parameter? I can look harder if you can't find something easily.